File: src/dump.rs
//! Hex dump utilities. Two modes:
//!   - raw: classic hex+ASCII view of any byte range
//!   - record-aware: parses Type Word + IRIG timestamp + Command Word and
//!     prints an annotated header per record before the bytes
//!
//! Mirrors the Python `dump.py` output format closely enough for diffing.

use std::fs::File;
use std::io::{BufWriter, Read, Write};
use std::path::Path;

use crate::decode::{
    MIN_RECORD_BYTES, MIN_RECORD_WORDS, decode_command_word, decode_irig_timestamp,
    decode_type_word, read_u16,
};
use crate::error::{MieError, MieResult};
use crate::models::{Direction, MessageType};

fn type_name(code: u8) -> String {
    match MessageType::from_code(code) {
        Some(MessageType::ModeCommand) => "Mode Command".into(),
        Some(MessageType::BcToRt) => "BC->RT (Receive)".into(),
        Some(MessageType::RtToBc) => "RT->BC (Transmit)".into(),
        Some(MessageType::RtToRt) => "RT->RT".into(),
        Some(MessageType::BroadcastBcToRt) => "Broadcast BC->RT".into(),
        Some(MessageType::BroadcastRtToRt) => "Broadcast RT->RT".into(),
        Some(MessageType::SpuriousData) => "Spurious Data".into(),
        None => format!("UNKNOWN(0x{code:02X})"),
    }
}

fn read_file(path: &Path) -> MieResult<Vec<u8>> {
    if !path.exists() {
        return Err(MieError::FileNotFound {
            path: path.to_path_buf(),
        });
    }
    let mut file = File::open(path).map_err(|source| MieError::FileIo {
        path: path.to_path_buf(),
        source,
    })?;
    let mut buf = Vec::new();
    file.read_to_end(&mut buf).map_err(|source| MieError::FileIo {
        path: path.to_path_buf(),
        source,
    })?;
    if buf.is_empty() {
        return Err(MieError::FileEmpty {
            path: path.to_path_buf(),
        });
    }
    Ok(buf)
}

fn write_hex_line<W: Write>(out: &mut W, indent: &str, addr: usize, chunk: &[u8]) -> std::io::Result<()> {
    write!(out, "{indent}{addr:08X}  ")?;
    let mut hex_part = String::with_capacity(48);
    for b in chunk {
        hex_part.push_str(&format!("{b:02X} "));
    }
    if hex_part.ends_with(' ') {
        hex_part.pop();
    }
    write!(out, "{hex_part:<48}  |")?;
    for &b in chunk {
        let c = if (32..127).contains(&b) { b as char } else { '.' };
        out.write_all(&[c as u8])?;
    }
    out.write_all(b"|\n")
}

/// Print a raw hex+ASCII dump of `[start_offset, start_offset + length)`.
pub fn hex_dump_raw(
    path: &Path,
    start_offset: usize,
    length: Option<usize>,
    mut out: impl Write,
) -> MieResult<()> {
    let data = read_file(path)?;
    let end = match length {
        Some(n) => (start_offset + n).min(data.len()),
        None => data.len(),
    };
    let chunk = &data[start_offset.min(data.len())..end];

    let name = path.file_name().map(|n| n.to_string_lossy().into_owned()).unwrap_or_default();
    let _ = writeln!(out, "File: {name} ({} bytes)", data.len());
    let _ = writeln!(out, "Range: 0x{:08X}-0x{:08X}", start_offset, end);
    let _ = writeln!(out);

    let mut i = 0;
    while i < chunk.len() {
        let line = &chunk[i..(i + 16).min(chunk.len())];
        let _ = write_hex_line(&mut out, "  ", start_offset + i, line);
        i += 16;
    }
    Ok(())
}

/// Print a record-aware hex dump. Each record is preceded by a one-line
/// header summarising decoded Type Word, timestamp, and Command Word.
pub fn hex_dump_records(
    path: &Path,
    max_records: Option<u64>,
    start_offset: usize,
    mut out: impl Write,
) -> MieResult<()> {
    let data = read_file(path)?;
    let file_len = data.len();
    let mut offset = start_offset;
    let mut record_num: u64 = 0;

    let name = path.file_name().map(|n| n.to_string_lossy().into_owned()).unwrap_or_default();
    let _ = writeln!(out, "File: {name} ({file_len} bytes)");
    let _ = writeln!(out, "Record dump starting at offset 0x{:08X}", start_offset);
    let _ = writeln!(out);

    while offset + MIN_RECORD_BYTES <= file_len {
        if let Some(max) = max_records {
            if record_num >= max {
                break;
            }
        }

        let type_raw = match read_u16(&data, offset) {
            Some(v) => v,
            None => break,
        };
        let tw = decode_type_word(type_raw);

        if tw.word_count < MIN_RECORD_WORDS {
            let _ = writeln!(
                out,
                "  !! Invalid word_count={} at 0x{:08X}, stopping",
                tw.word_count, offset
            );
            break;
        }
        let record_bytes = usize::from(tw.word_count) * 2;
        if offset + record_bytes > file_len {
            let _ = writeln!(
                out,
                "  !! Truncated record at 0x{:08X} ({} bytes needed, {} available)",
                offset, record_bytes, file_len - offset
            );
            break;
        }

        // Decode IRIG-shaped header for annotation. (For Standard timestamps
        // this is still useful to display the raw bytes; the IRIG decode is
        // a best-effort summary.)
        let ts_upper = read_u16(&data, offset + 2).unwrap_or(0);
        let ts_middle = read_u16(&data, offset + 4).unwrap_or(0);
        let ts_lower = read_u16(&data, offset + 6).unwrap_or(0);
        let ts = decode_irig_timestamp(ts_upper, ts_middle, ts_lower);
        let cmd_raw = read_u16(&data, offset + 8).unwrap_or(0);
        let cmd = decode_command_word(cmd_raw);

        let bus_label = tw.bus.as_str();
        let err_label = if tw.error { "ERROR" } else { "OK" };
        let dir_char = if cmd.direction == Direction::Transmit {
            'T'
        } else {
            'R'
        };

        let _ = writeln!(out, "{}", "-".repeat(72));
        let _ = writeln!(
            out,
            "  Record #{record_num}  @  0x{offset:08X}  ({record_bytes} bytes, {wc} words)",
            wc = tw.word_count
        );
        let _ = writeln!(
            out,
            "  Type: 0x{:04X}  ->  {}  Bus {}  {}",
            tw.raw,
            type_name(tw.message_type),
            bus_label,
            err_label
        );
        let _ = writeln!(
            out,
            "  Time: {}{}",
            ts.format(),
            if ts.freerun { "  [FREERUN]" } else { "" }
        );
        let _ = writeln!(
            out,
            "  Cmd:  0x{:04X}  ->  RT{} SA{} {} WC={}",
            cmd.raw, cmd.rt, cmd.subaddress, dir_char, cmd.data_word_count
        );

        let record_data = &data[offset..offset + record_bytes];
        let mut i = 0;
        while i < record_data.len() {
            let line = &record_data[i..(i + 16).min(record_data.len())];
            let _ = write_hex_line(&mut out, "    ", offset + i, line);
            i += 16;
        }
        let _ = writeln!(out);

        offset += record_bytes;
        record_num += 1;
    }

    let _ = writeln!(out, "{}", "-".repeat(72));
    let _ = writeln!(out, "{record_num} records dumped.");
    Ok(())
}

/// Convenience wrapper that writes to stdout (buffered).
pub fn hex_dump_raw_to_stdout(path: &Path, offset: usize, length: Option<usize>) -> MieResult<()> {
    let stdout = std::io::stdout();
    let buf = BufWriter::new(stdout.lock());
    hex_dump_raw(path, offset, length, buf)
}

pub fn hex_dump_records_to_stdout(
    path: &Path,
    max_records: Option<u64>,
    offset: usize,
) -> MieResult<()> {
    let stdout = std::io::stdout();
    let buf = BufWriter::new(stdout.lock());
    hex_dump_records(path, max_records, offset, buf)
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::atomic::{AtomicU64, Ordering};

    struct TempFile(std::path::PathBuf);
    impl TempFile {
        fn write(bytes: &[u8]) -> Self {
            static C: AtomicU64 = AtomicU64::new(0);
            let n = C.fetch_add(1, Ordering::Relaxed);
            let pid = std::process::id();
            let p = std::env::temp_dir().join(format!("mie-dump-test-{pid}-{n}.bin"));
            std::fs::write(&p, bytes).unwrap();
            Self(p)
        }
        fn path(&self) -> &Path {
            &self.0
        }
    }
    impl Drop for TempFile {
        fn drop(&mut self) {
            let _ = std::fs::remove_file(&self.0);
        }
    }

    #[test]
    fn raw_dump_format() {
        let f = TempFile::write(b"AB\x00\x01\x02\x03");
        let mut out = Vec::new();
        hex_dump_raw(f.path(), 0, None, &mut out).unwrap();
        let s = String::from_utf8(out).unwrap();
        assert!(s.contains("00000000"));
        assert!(s.contains("41 42")); // "AB"
        assert!(s.contains("|AB"));
    }

    #[test]
    fn record_dump_handles_one_record() {
        // Minimal valid record: type 0x2402, then 35 zero words = 70 bytes
        let mut buf = Vec::with_capacity(72);
        buf.extend_from_slice(&[0x02, 0x24]);
        buf.extend_from_slice(&[0u8; 70]);
        let f = TempFile::write(&buf);
        let mut out = Vec::new();
        hex_dump_records(f.path(), Some(1), 0, &mut out).unwrap();
        let s = String::from_utf8(out).unwrap();
        assert!(s.contains("Record #0"));
        assert!(s.contains("BC->RT (Receive)"));
        assert!(s.contains("1 records dumped"));
    }

    #[test]
    fn missing_file_returns_error() {
        let mut out = Vec::new();
        let err = hex_dump_raw(Path::new("/no/such/file.bin"), 0, None, &mut out).unwrap_err();
        assert_eq!(err.kind(), crate::error::MieErrorKind::FileNotFound);
    }
}

File: cpp/src/dump.cpp
// SPDX-License-Identifier: Apache-2.0

#define MIE_LOG_MODULE "mie_decoder::dump"

#include "mie/dump.hpp"

#include <string>
#include <vector>

#include "mie/decode.hpp"
#include "mie/error.hpp"
#include "mie/log.hpp"
#include "mie/models.hpp"
#include "mie/platform.hpp"
#include "mie/text.hpp"

namespace mie {
namespace dump {

namespace {

const std::size_t kBytesPerLine = 16;
/// Width of the hex column: 16 bytes as "XX " minus the trailing space.
const std::size_t kHexColumnWidth = 48;
const std::size_t kRuleWidth = 72;

/// Human label for a message type.
///
/// Deliberately NOT `models::message_type_name`, which returns the enum-style
/// `BC_TO_RT` used in the CSV and in error text. This view is read by a person
/// staring at a broken file, and the arrow form says which way the traffic went
/// without them having to remember the convention.
std::string type_label(uint8_t code) {
    switch (code) {
        case MESSAGE_TYPE_MODE_COMMAND: return "Mode Command";
        case MESSAGE_TYPE_BC_TO_RT: return "BC->RT (Receive)";
        case MESSAGE_TYPE_RT_TO_BC: return "RT->BC (Transmit)";
        case MESSAGE_TYPE_RT_TO_RT: return "RT->RT";
        case MESSAGE_TYPE_BROADCAST_BC_TO_RT: return "Broadcast BC->RT";
        case MESSAGE_TYPE_BROADCAST_RT_TO_RT: return "Broadcast RT->RT";
        case MESSAGE_TYPE_SPURIOUS_DATA: return "Spurious Data";
        default:
            // Named, not hidden. An unknown type code is the most interesting
            // thing on the line when a scan has gone wrong.
            return "UNKNOWN(0x" + text::hex_upper(code, 2) + ")";
    }
}

/// Canonical uppercase name of a classified message format.
///
/// Deliberately NOT `models::message_format_name`, which returns the CamelCase
/// spelling used elsewhere. The dump report is diffed against the Rust and
/// Python dumps, both of which print the SCREAMING_SNAKE form here -- so this
/// is contract, not preference, and the two spellings must not be conflated.
const char* format_label(MessageFormat format) {
    switch (format) {
        case FORMAT_RECEIVE: return "RECEIVE";
        case FORMAT_TRANSMIT: return "TRANSMIT";
        case FORMAT_RT_TO_RT: return "RT_TO_RT";
        case FORMAT_RECEIVE_BROADCAST: return "RECEIVE_BROADCAST";
        case FORMAT_RT_TO_RT_BROADCAST: return "RT_TO_RT_BROADCAST";
        case FORMAT_MODE_CODE_TX_DATA: return "MODE_CODE_TX_DATA";
        case FORMAT_MODE_CODE_RX_DATA: return "MODE_CODE_RX_DATA";
        case FORMAT_MODE_CODE_NO_DATA: return "MODE_CODE_NO_DATA";
        case FORMAT_MODE_CODE_BCAST_NO_DATA: return "MODE_CODE_BCAST_NO_DATA";
        case FORMAT_MODE_CODE_BCAST_DATA: return "MODE_CODE_BCAST_DATA";
        case FORMAT_SPURIOUS_DATA: return "SPURIOUS_DATA";
    }
    return "";
}

std::string rule() { return std::string(kRuleWidth, '-'); }

/// The last path component, or the whole string when there is no separator.
std::string file_name_of(const std::string& path) {
    const std::size_t slash = path.find_last_of("/\\");
    return slash == std::string::npos ? path : path.substr(slash + 1);
}

/// Read the whole file.
///
/// Read rather than mapped: this view must work on files the reader rejects,
/// and the ranges asked for are small and arbitrary rather than a walk.
std::vector<uint8_t> read_whole_file(const std::string& path) {
    // Through the platform layer: a recording may sit at a non-ASCII path, and
    // std::fopen would fail to find it on Windows.
    std::vector<uint8_t> data;
    platform::OsError err;
    if (!platform::read_file(path, data, err)) {
        // "not found" and "could not be read" are different problems with
        // different remedies, and a directory stats as zero bytes on Windows --
        // so the existence test decides which this was rather than the size.
        if (!platform::path_exists(path)) {
            throw MieError::file_not_found(path);
        }
        throw MieError::file_io(path, err.message, err.code);
    }
    if (data.empty()) {
        throw MieError::file_empty(path);
    }
    return data;
}

/// Write `text`, raising a writer error on failure.
///
/// Every write is checked. `mie-decoder dump x.mie | head` closes the pipe
/// mid-report, and a dump that ignored the failure would exit 0 having produced
/// a truncated view -- indistinguishable, to a script, from a short file.
void emit(std::FILE* out, const std::string& text) {
    if (std::fwrite(text.data(), 1, text.size(), out) != text.size()) {
        platform::OsError err;
        platform::capture_stream_error(err);
        throw MieError::writer_error("stdout", err.message, err.code);
    }
}

void emit_line(std::FILE* out, const std::string& text) { emit(out, text + "\n"); }

/// One hex+ASCII line: address, up to 16 bytes in hex, then the printable
/// rendering between pipes.
void write_hex_line(std::FILE* out, const char* indent, std::size_t address, const uint8_t* bytes,
                    std::size_t count) {
    std::string line = indent;
    line += text::hex_upper(address, 8);
    line += "  ";

    std::string hex;
    for (std::size_t i = 0; i < count; ++i) {
        if (i > 0) {
            hex += ' ';
        }
        hex += text::hex_upper(bytes[i], 2);
    }
    line += hex;
    // Padded so the ASCII column lines up even on a short final line.
    if (hex.size() < kHexColumnWidth) {
        line.append(kHexColumnWidth - hex.size(), ' ');
    }
    line += "  |";
    for (std::size_t i = 0; i < count; ++i) {
        // Explicit ASCII range, never <cctype>: this tree is locale-free by
        // rule (scripts/assert-locale-free.sh), and `isprint` under a non-C
        // locale would render high bytes differently on different hosts.
        const uint8_t byte = bytes[i];
        line += (byte >= 32 && byte < 127) ? static_cast<char>(byte) : '.';
    }
    line += "|";
    emit_line(out, line);
}

void write_hex_block(std::FILE* out, const char* indent, const uint8_t* data, std::size_t size,
                     std::size_t base_address) {
    for (std::size_t i = 0; i < size; i += kBytesPerLine) {
        const std::size_t count = (size - i < kBytesPerLine) ? (size - i) : kBytesPerLine;
        write_hex_line(out, indent, base_address + i, data + i, count);
    }
}

/// Validate the record at `offset`. False stops the scan, having written the
/// anomaly inline and logged it (L2-CLI-013).
bool record_extent(const TypeWord& type_word, std::size_t offset, std::size_t file_size,
                   std::FILE* out, std::size_t& record_end) {
    if (type_word.word_count < decode::MIN_RECORD_WORDS) {
        const std::string where = text::hex_upper(offset, 8);
        emit_line(out, "  !! Invalid word_count=" + text::decimal(type_word.word_count) + " at 0x" +
                           where + ", stopping");
        MIE_LOG_WARN("dump: invalid word_count=" + text::decimal(type_word.word_count) + " at 0x" +
                     text::hex_upper(offset, 1) + "; stopping record scan");
        return false;
    }

    const std::size_t record_bytes = static_cast<std::size_t>(type_word.word_count) * 2;
    // Belt and braces. `word_count` is six bits, so `record_bytes` cannot exceed
    // 126 and this cannot fire for an offset inside a real file -- but the
    // alternative to checking is silently wrapping to a small offset and
    // dumping the wrong bytes as though they were right.
    if (offset > static_cast<std::size_t>(-1) - record_bytes) {
        emit_line(out, "  !! Offset overflow at 0x" + text::hex_upper(offset, 8) +
                           " (record_bytes=" + text::decimal(record_bytes) + "), stopping");
        MIE_LOG_WARN("dump: offset overflow at 0x" + text::hex_upper(offset, 1) +
                     " (record_bytes=" + text::decimal(record_bytes) + "); stopping record scan");
        return false;
    }

    const std::size_t end = offset + record_bytes;
    if (end > file_size) {
        emit_line(out, "  !! Truncated record at 0x" + text::hex_upper(offset, 8) + " (" +
                           text::decimal(record_bytes) + " bytes needed, " +
                           text::decimal(file_size - offset) + " available)");
        MIE_LOG_WARN("dump: truncated record at 0x" + text::hex_upper(offset, 1) + " (" +
                     text::decimal(record_bytes) + " bytes needed, " +
                     text::decimal(file_size - offset) + " available); stopping record scan");
        return false;
    }
    record_end = end;
    return true;
}

uint16_t word_at(const std::vector<uint8_t>& data, std::size_t offset) {
    uint16_t value = 0;
    if (!decode::read_u16(data.empty() ? NULL : &data[0], data.size(), offset, value)) {
        return 0;
    }
    return value;
}

/// The decoded-header block for one record.
void write_annotation(std::FILE* out, const std::vector<uint8_t>& data, const TypeWord& type_word,
                      std::size_t offset, std::size_t record_bytes, uint64_t record_number) {
    // The timestamp is decoded as IRIG unconditionally. For a Standard-format
    // file this line is wrong and the raw bytes below remain authoritative --
    // which is the honest trade for a view whose whole job is to run before the
    // format has been established.
    const IrigTimestamp timestamp = decode::decode_irig_timestamp(
        word_at(data, offset + 2), word_at(data, offset + 4), word_at(data, offset + 6));
    const CommandWord command = decode::decode_command_word(word_at(data, offset + 8));

    MessageFormat format = FORMAT_RECEIVE;
    // A record that cannot be classified gets a label, not an error: this runs
    // on suspect files by definition.
    const bool classified = decode::classify_message_format(type_word.message_type, command,
                                                            type_word.word_count, 3, format);

    emit_line(out, rule());
    emit_line(out, "  Record #" + text::decimal(record_number) + "  @  0x" +
                       text::hex_upper(offset, 8) + "  (" + text::decimal(record_bytes) +
                       " bytes, " + text::decimal(type_word.word_count) + " words)");
    emit_line(out, "  Type:   0x" + text::hex_upper(type_word.raw, 4) + "  ->  " +
                       type_label(type_word.message_type) + "  Bus " + bus_name(type_word.bus) +
                       "  error flag (bit 14): " + (type_word.error ? "SET" : "clear"));
    emit_line(out,
              std::string("  Format: ") + (classified ? format_label(format) : "(unclassifiable)"));
    emit_line(out, "  Time:   " + timestamp.format() + (timestamp.freerun ? "  [FREERUN]" : ""));
    emit_line(out, "  Cmd:    0x" + text::hex_upper(command.raw, 4) + "  ->  RT" +
                       text::decimal(command.rt) + " SA" + text::decimal(command.subaddress) + " " +
                       (command.direction == DIRECTION_TRANSMIT ? "T" : "R") +
                       " WC=" + text::decimal(command.data_word_count));

    if (type_word.error) {
        // The Error Word is the last word of an errored record. Showing its DDC
        // description here is what makes the reason legible without a trip to
        // the error catalogue.
        const uint16_t code =
            word_at(data, offset + (static_cast<std::size_t>(type_word.word_count) - 1) * 2);
        emit_line(out, "  Error:  0x" + text::hex_upper(code, 4) + "  ->  " +
                           ddc_error_description_or_unknown(code));
    }
}

}  // namespace

void hex_dump_raw(const std::string& path, std::size_t offset, const Optional<std::size_t>& length,
                  std::FILE* out) {
    const std::vector<uint8_t> data = read_whole_file(path);
    const std::size_t file_size = data.size();

    // Clamped, not validated. An operator exploring a damaged file should get
    // an empty range back rather than an argument error -- the empty output
    // already says the offset is past the end.
    const std::size_t start = offset < file_size ? offset : file_size;
    std::size_t end = file_size;
    if (length.has_value()) {
        const std::size_t requested = length.value();
        // Saturate rather than wrap: `--offset 4096 --length <huge>` must mean
        // "to the end", not "back to the beginning".
        end = (offset > static_cast<std::size_t>(-1) - requested) ? file_size : offset + requested;
        if (end > file_size) {
            end = file_size;
        }
    }
    if (end < start) {
        end = start;
    }

    emit_line(out, "File: " + file_name_of(path) + " (" + text::decimal(file_size) + " bytes)");
    emit_line(out, "Range: 0x" + text::hex_upper(start, 8) + "-0x" + text::hex_upper(end, 8));
    emit_line(out, "");

    write_hex_block(out, "  ", &data[0] + start, end - start, start);
    if (std::fflush(out) != 0) {
        platform::OsError err;
        platform::capture_stream_error(err);
        throw MieError::writer_error("stdout", err.message, err.code);
    }
}

void hex_dump_records(const std::string& path, const Optional<uint64_t>& max_records,
                      std::size_t offset, std::FILE* out) {
    const std::vector<uint8_t> data = read_whole_file(path);
    const std::size_t file_size = data.size();

    emit_line(out, "File: " + file_name_of(path) + " (" + text::decimal(file_size) + " bytes)");
    emit_line(out, "Record dump starting at offset 0x" + text::hex_upper(offset, 8));
    emit_line(out, "");

    std::size_t at = offset;
    uint64_t record_number = 0;
    for (;;) {
        if (at > static_cast<std::size_t>(-1) - decode::MIN_RECORD_BYTES) {
            break;
        }
        if (at + decode::MIN_RECORD_BYTES > file_size) {
            break;
        }
        if (max_records.has_value() && record_number >= max_records.value()) {
            break;
        }

        uint16_t type_raw = 0;
        if (!decode::read_u16(&data[0], file_size, at, type_raw)) {
            break;
        }
        const TypeWord type_word = decode::decode_type_word(type_raw);

        std::size_t record_end = 0;
        if (!record_extent(type_word, at, file_size, out, record_end)) {
            break;
        }

        write_annotation(out, data, type_word, at, record_end - at, record_number);
        write_hex_block(out, "    ", &data[0] + at, record_end - at, at);
        emit_line(out, "");

        at = record_end;
        record_number += 1;
    }

    emit_line(out, rule());
    emit_line(out, text::decimal(record_number) + " records dumped.");
    if (std::fflush(out) != 0) {
        platform::OsError err;
        platform::capture_stream_error(err);
        throw MieError::writer_error("stdout", err.message, err.code);
    }
}

}  // namespace dump
}  // namespace mie
